iceoryx_dds/include/iceoryx_dds/internal/gateway/dds_gateway_generic.inl

```diff
@@ -0,0 +1,160 @@
+    {
+        iox::dds::LogDebug() << "[DDSGatewayGeneric] Shutting down Posh2DDSGateway.";
+
+        m_isRunning.store(false, std::memory_order_relaxed);
+
+        m_discoveryThread.join();
+        m_forwardingThread.join();
+    }
+}
+
+template <typename channel_t, typename gateway_t>
+inline uint64_t iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::getNumberOfChannels() const noexcept
+{
+    auto guardedVector = m_channels.GetScopeGuard();
+    return guardedVector->size();
+}
+
+// ================================================== Protected ================================================== //
+
+template <typename channel_t, typename gateway_t>
+inline iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::DDSGatewayGeneric() noexcept
+    : gateway_t(iox::capro::Interfaces::DDS)
+{
+    LogDebug() << "[DDSGatewayGeneric] Using default channel factory.";
+    m_channelFactory = channel_t::create;
+}
+
+template <typename channel_t, typename gateway_t>
+inline channel_t
+iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::addChannel(const iox::capro::ServiceDescription& service) noexcept
+{
+    auto channel = m_channelFactory(service);
+    m_channels->push_back(channel);
+    iox::dds::LogDebug() << "[DDSGatewayGeneric] Channel set up for service: "
+                         << "/" << service.getInstanceIDString() << "/" << service.getServiceIDString() << "/"
+                         << service.getEventIDString();
+    return channel;
+}
+
+template <typename channel_t, typename gateway_t>
+inline iox::cxx::optional<channel_t>
+iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::findChannel(const iox::capro::ServiceDescription& service) const
+    noexcept
+{
+    auto guardedVector = this->m_channels.GetScopeGuard();
+    auto channel = std::find_if(guardedVector->begin(), guardedVector->end(), [&service](const channel_t& channel) {
+        return channel.getService() == service;
+    });
+    if (channel == guardedVector->end())
+    {
+        return iox::cxx::nullopt_t();
+    }
+    else
+    {
+        // Needs to be dereferenced since the smart lock returns pointers to elements in concurrent collections.
+        return iox::cxx::make_optional<channel_t>(*channel);
+    }
+}
+
+template <typename channel_t, typename gateway_t>
+inline void
+iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::forEachChannel(const std::function<void(channel_t&)> f) const
+    noexcept
+{
+    auto guardedVector = m_channels.GetScopeGuard();
+    for (auto channel = guardedVector->begin(); channel != guardedVector->end(); ++channel)
+    {
+        f(*channel);
+    }
+}
+
+template <typename channel_t, typename gateway_t>
+inline void iox::dds::DDSGatewayGeneric<channel_t, gateway_t>::discardChannel(
+    const iox::capro::ServiceDescription& service) noexcept
+{
+    auto guardedVector = this->m_channels.GetScopeGuard();
+    auto channel = std::find_if(guardedVector->begin(), guardedVector->end(), [&service](const channel_t& channel) {
```

Comment:
have you thought about using std::remove_if?

Reply:
:heavy_check_mark:

Reply:
Nevermind, the following code does not work for some reason... 
```
    auto guardedVector = this->m_channels.GetScopeGuard();
    std::remove_if(guardedVector->begin(), guardedVector->end(), [&service](const channel_t& channel) {
        return channel.getService() == service;
    });
```
Will leave it as it was for now.